src/main/java/org/openlmis/buq/service/buq/BottomUpQuantificationService.java

```diff
@@ -0,0 +1,232 @@
+  }
+
+  private void addNewStatusChange(BottomUpQuantification bottomUpQuantification) {
+    bottomUpQuantification.getStatusChanges().add(
+        BottomUpQuantificationStatusChange.newInstance(
+            bottomUpQuantification,
+            authenticationHelper.getCurrentUser().getId(),
+            bottomUpQuantification.getStatus())
+    );
+  }
+
+  private BottomUpQuantification updateBottomUpQuantification(
+      BottomUpQuantificationDto bottomUpQuantificationDto, UUID bottomUpQuantificationId) {
+    if (null != bottomUpQuantificationDto.getId()
+        && !Objects.equals(bottomUpQuantificationDto.getId(), bottomUpQuantificationId)) {
+      throw new ValidationMessageException(ERROR_ID_MISMATCH);
+    }
+
+    BottomUpQuantification bottomUpQuantificationToUpdate =
+        findBottomUpQuantification(bottomUpQuantificationId);
+
+    BottomUpQuantification updatedBottomUpQuantification = BottomUpQuantification
+        .newInstance(bottomUpQuantificationDto);
+    updatedBottomUpQuantification.setId(bottomUpQuantificationId);
+
+    List<BottomUpQuantificationLineItem> updatedLineItems = bottomUpQuantificationDto
+        .getBottomUpQuantificationLineItems()
+        .stream()
+        .map(lineItemDto -> {
+          BottomUpQuantificationLineItem lineItem = BottomUpQuantificationLineItem
+              .newInstance(lineItemDto);
+          lineItem.setBottomUpQuantification(bottomUpQuantificationToUpdate);
+          lineItem.setId(lineItemDto.getId());
+
+          return lineItem;
+        })
+        .collect(Collectors.toList());
+    bottomUpQuantificationToUpdate.updateFrom(updatedBottomUpQuantification, updatedLineItems);
+
+    return bottomUpQuantificationToUpdate;
+  }
+
+  private FacilityDto findFacility(UUID facilityId) {
+    return findResource(facilityId, facilityReferenceDataService::findOne,
+        ERROR_FACILITY_NOT_FOUND);
+  }
+
+  private ProgramDto findProgram(UUID programId) {
+    return findResource(programId, programReferenceDataService::findOne,
+        ERROR_PROGRAM_NOT_FOUND);
+  }
+
+  private ProcessingPeriodDto findPeriod(UUID programId) {
+    return findResource(programId, periodReferenceDataService::findOne,
+        ERROR_PROCESSING_PERIOD_NOT_FOUND);
+  }
+
+  private <R> R findResource(UUID id, Function<UUID, R> finder, String errorMessage) {
+    return Optional
+        .ofNullable(finder.apply(id))
+        .orElseThrow(() -> {
+          return new ContentNotFoundMessageException(errorMessage, id);
+        });
+  }
+
+  BottomUpQuantification findBottomUpQuantification(UUID bottomUpQuantificationId) {
+    BottomUpQuantification bottomUpQuantification = bottomUpQuantificationRepository
+        .findById(bottomUpQuantificationId).orElse(null);
+    if (null == bottomUpQuantification) {
+      throw new ContentNotFoundMessageException(ERROR_BOTTOM_UP_QUANTIFICATION_NOT_FOUND,
+          bottomUpQuantificationId);
+    }
+
+    return bottomUpQuantification;
```

Comment:
wouldn't it be better to shorten it and use `.findById(...).orElseThrow(...)`?

Reply:
I changed the method to use orElseThrow()